tools/buildsys/src/builder.rs

```diff
@@ -411,34 +419,74 @@ where
 /// Remove build artifacts from the output directory.
 /// Any marker file we find could have a corresponding file that should be cleaned up.
 /// We also clean up the marker files so they do not accumulate across builds.
+/// For the same reason, if a directory is empty after build artifacts, marker files, and other
+/// empty directories have been removed, then that directory will also be removed.
 fn clean_build_files<P>(build_dir: P, output_dir: P) -> Result<()>
 where
     P: AsRef<Path>,
 {
+    let build_dir = build_dir.as_ref();
+    let output_dir = output_dir.as_ref();
+
     fn is_marker(entry: &DirEntry) -> bool {
-        entry
-            .file_name()
-            .to_str()
-            .map(|s| s.ends_with(MARKER_EXTENSION))
-            .unwrap_or(false)
+        entry.path().is_dir()
+            || entry
+                .file_name()
+                .to_str()
+                .map(|s| s.ends_with(MARKER_EXTENSION))
+                .unwrap_or(false)
     }
 
-    for marker_file in find_files(&build_dir, is_marker) {
-        let mut output_file: PathBuf = output_dir.as_ref().into();
-        output_file.push(
-            marker_file
-                .file_name()
-                .context(error::BadFilenameSnafu { path: &marker_file })?,
-        );
+    fn cleanup(path: &Path, top: &Path, dirs: &mut HashSet<PathBuf>) -> Result<()> {
+        if !path.exists() {
+            return Ok(());
+        }
+        std::fs::remove_file(&path).context(error::FileRemoveSnafu { path })?;
+        let mut parent = path.parent();
+        while let Some(p) = parent {
+            if p == top || dirs.contains(p) {
+                break;
+            }
+            dirs.insert(p.into());
+            parent = p.parent()
+        }
```

Comment:
Will it be possible to clean up the parent directories here as well? Or, why do they have to be deleted after the marker and output file are deleted?

Reply:
It would be possible but not as efficient. For `N` output files, we'd end up checking whether the parent directory was empty `N` times, rather than once.